### ISSUE TITLE
2장 loss값 계산 관련

### ISSUE THREAD

**Issue by sin222** [Justin-A/DeepLearning101#14]
2장뿐만이 아니라 전체적으로 loss값 계산할 때 보면 len(test_loader.dataset)으로 나누게 되는데 loss값은 len(test_loader)로 나눠야 맞는게 아닌건가 싶어 여쭤봅니다. 감사합니다.
예제 코드는 아래와 같습니다.

''' 9. 학습되는 과정 속에서 검증 데이터에 대한 모델 성능을 확인하는 함수 정의 '''
def evaluate(model, test_loader):
    model.eval()
    test_loss = 0
    correct = 0

    with torch.no_grad():
        for image, label in test_loader:
            image = image.to(DEVICE)
            label = label.to(DEVICE)
            output = model(image)
            test_loss += criterion(output, label).item() #이부분입니다.
            prediction = output.max(1, keepdim = True)[1]
            correct += prediction.eq(label.view_as(prediction)).sum().item()
    
    test_loss /= len(test_loader.dataset) #이부분입니다.
    test_accuracy = 100. * correct / len(test_loader.dataset)
    return test_loss, test_accuracy

**Comment by Justin-A** [comment]
안녕하세요, Justin입니다.

문의해주신 내용을 확인해본 결과, len(test_loader)일 경우, test_loader는 배치 단위로 데이터를 구성하고 있기 때문에 배치 개수로 계산이 됩니다. 하지만 저희는 loss를 계산할 때, 배치 단위가 아니라 이미지 1장에 대해서 loss값을 계산하고 이를 평균을 계산해야 하기 때문에 len(test_loader.dataset)을 진행해야 합니다.

자세한 내용은 PyTorch의 data_loader 부분을 참고하시면 될 것 같습니다.
참고 링크 : https://pytorch.org/docs/stable/data.html

감사합니다.

**Comment by sin222** [comment]
답변 감사합니다. 몇 가지 이해가 안돼 더 질문드립니다.
이미지가 100개가 있고 batch_size가 10일 경우 len(test_loader)는 10이 돼고 그렇기 때문에 저 for문은 10번 반복한다는 것을 확인했습니다. 
그래서 test_loss도 총 10개의 값이 합쳐진 상태였는데 for문이 끝나고나서는 이미지의 개수인 1000개로 나눈다는게 이해가 잘 가지 않습니다.

**Comment by Justin-A** [comment]
loss를 구한다는 개념은 데이터 1개에 대한 값을 구하는 것이며, 배치 단위로 진행될 경우 배치 내 포함된 데이터 별 loss의 평균을 계산한다는 의미입니다. 
예로 들어주신 batch_size 10일 경우, 1개의 배치에 속한 데이터 10개에 대한, 즉 이미지 10장에 대한 loss를 계산하겠죠? 
그렇다면 10개의 loss값이 산출되는데 이의 평균값을 이용하여 1개의 배치의 로스를 계산합니다. 
이 배치의 loss값을 이용하여 파라미터를 업데이트 하는 back propagation이 이뤄지는 것이구요. 
testset의 경우 모델이 학습되는 것이 아니기 때문에 배치 별 loss를 구할 필요가 없겠죠? 
testset의 loss는 단지 전체 데이터의 loss 평균을 확인하는 용도입니다.

**Comment by sin222** [comment]
산출된 loss값이 mini_batch에 속한 10개 이미지 loss의 합이 아닌 평균값이기 때문에 코드상 test_loss값은 10번 더해지는 것이고 그렇기 때문에 전체 데이터의 평균도 test_loss를 미니배치 개수인 10개만큼 나눠야 하는 것은 아닌지 궁금합니다!

100으로 나누려면 이미지 한개당 test_loss값이(즉, 각각 구한 loss값 100개) 모두 더해져야 하는 것 아닌가요..?,,ㅠㅠ

빠른답변 너무너무 감사합니다 :)

**Comment by Justin-A** [comment]
test_loss값을 이미지별로 계산한 값을 모두 더해주고 이를 전체 데이터 개수로 나누어주는 것이기 때문에 배치 개수와는 무관합니다.

**Comment by tjqansthd** [comment]
안녕하세요, 책의 예제에서는 ``train_loader``와 ``test_loader``가 선언되는 코드에서 ``batch_size`` 인자로 1이 들어가지 않고, ``train_loader``와 동일하게 ``BATCH_SIZE``가 들어갑니다. 그래서 ``BATCH_SIZE``를 다르게 설정하면 ``test_loader``의 ``len(test_loader)`` 또한 바뀌고, evaluate 함수 내에서 for문의 iteration 수 또한 바뀌기 때문에 그런 것 같습니다. 

``BATCH_SIZE=10``이라면 반복문에서 이미지 '10장'에 대한 평균 loss값이 전체이미지수/10 만큼의 횟수로 더해지게 됩니다. loss는 10장에 대한 평균을 내던 100장에 대한 평균을 내던 같은 스케일을 가질 텐데, iteration수는 바뀌기 때문에 더해진 총 값의 크기가 BATCH_SIZE에 따라 바뀝니다. ``len(test_loader.dataset)``으로 나누려면 test_loader의 전달인자를 넣는 부분에서 인자를 batch_size=1 로 주는게 맞는 것 같습니다. 

혹은, ``len(test_loader)``는 ``batch_size=1``일때 ``len(test_loader.dataset)``과 같은 값을 가지므로 오히려 ``test_loader``의 ``batch_size``에 ``BATCH_SIZE`` 만큼을 할당하고 ``test_loss /= len(test_loader)`` 와 같이 로스를 계산하게 하는 것이 더 일반성을 고려하는 코드인 것 같습니다.

**Comment by kalibianw** [comment]
안녕하세요, 저 역시 동일한 문제로 issue 남기게 되었습니다.

저 역시 @tjqansthd 님과 동일하게 생각합니다.

p.136의 코드를 예시로 들면 
```
def evaluate(model, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for image, label in test_loader:
            image = image.to(DEVICE)
            label = label.to(DEVICE)
            output = model(image)
            test_loss += criterion(output, label).item()
            prediction = output.max(1, keepdim=True)[1]
            correct += prediction.eq(label.view_as(prediction)).sum().item()
    test_loss /= len(test_loader.dataset)
    test_accuracy = 100. * correct / len(test_loader.dataset)
    
    return test_accuracy, test_loss
```
이 부분에서 `test_loss` 변수에 `criterion(output, label).item()`의 값을 모두 저장한 후 `len(test_loader.dataset)`로 나눠주는 부분이 있는데, 이 부분을 `test_loss /= (len(test_loader.dataset) / BATCH_SIZE)`로 바꾸면 loss값이 Epoch 한 번의 평균 값으로 출력이 잘 됩니다.

`criterion = nn.CrossEntropyLoss()`를 수행할 때 `reduction` argument의 기본 값이 `"mean"`이므로 해당 Batch에 Loss의 평균 값이 리턴되는 것으로 이해했는데(https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html 를 참고했습니다.), 특별히 다른 값을 지정하지 않고 사용하였으니 `test_loss /= (len(test_loader.dataset) / BATCH_SIZE)` 이처럼 나눠주는 것이 맞다고 생각합니다.
